lib/spack/spack/util/package_hash.py

```diff
@@ -74,82 +81,206 @@ def visit_ClassDef(self, node):  # noqa
+
+    If we're at version 1.0, we know that implementation 2 will win, because it
+    overrides implementation 1.  We should only include implementation 2.
+
+    If we're at version 3.0, we know that implementation 4 will win, because it
+    overrides implementation 1 (the default), and some @when on all others will be
+    False.
+
+    If we're at version 2.0, it's a bit more complicated. We know we can remove
+    implementations 2 and 4, because their @when's will never be satisfied. But, the
+    choice between implementations 1 and 3 will happen at runtime (this is a bad example
+    because the spec itself has platform information, and we should prefer to use that,
+    but we allow arbitrary boolean expressions in @when's, so this example suffices).
+    For this case, we end up needing to include *both* implementation 1 and 3 in the
+    package hash, because either could be chosen.
+
+    """
     def __init__(self, methods):
         self.methods = methods
 
-    def resolve(self, node):
-        if node.name not in self.methods:
-            raise PackageHashError(
-                "Future traversal visited new node: %s" % node.name)
-
-        result = None
-        for n, cond in self.methods[node.name]:
-            if cond:
-                return n
-            if cond is None:
-                result = n
+    def resolve(self, impl_conditions):
+        """Given list of nodes and conditions, figure out which node will be chosen."""
+        result = []
+        default = None
+        for impl, conditions in impl_conditions:
+            # if there's a default implementation with no conditions, remember that.
+            if not conditions:
+                default = impl
+                result.append(default)
+                continue
+
+            # any known-false @when means the method won't be used
+            if any(c is False for c in conditions):
+                continue
+
+            # anything with all known-true conditions will be picked if it's first
+            if all(c is True for c in conditions):
+                if result and result[0] is default:
+                    return [impl]  # we know the first MM will always win
+                # if anything dynamic comes before it we don't know if it'll win,
+                # so just let this result get appended
+
+            # anything else has to be determined dynamically, so add it to a list
+            result.append(impl)
+
+        # if nothing was picked, the last definition wins.
         return result
 
-    def visit_FunctionDef(self, node):  # noqa
-        if self.resolve(node) is node:
-            node.decorator_list = []
-            return node
-        return None
+    def visit_FunctionDef(self, func):  # noqa
+        # if the function def wasn't visited on the first traversal there is a problem
+        assert func.name in self.methods, "Inconsistent package traversal!"
```

Comment:
Recommend turning this into an `Exception` subclass before merging.

Reply:
Why?  If this isn't true, it's a logic error that should come out in testing, not something that should be dealt with at runtime by any client code.

Reply:
Ok, I misunderstood the intent of "there is a problem" in the comment! I think this is fine as-is then.